Core-collapse Supernova Profile Readin

pajkosmi opened the pull request:
## Proposed changes

This change reads in initial data from a supernova progenitor file and allows the star to collapse with a fixed metric after providing user-defined polytropic EOS quantities.

<!--
At a high level, describe what this PR does.
-->

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

pajkosmi committed:
Add analytic data for CCSN profile read from disk

Create CCSNCollapse readin capabilities

Add 4/3 polytrope data file

Include Ccsn test

nilsdeppe commented on src/PointwiseFunctions/AnalyticData/GrMhd/CcsnCollapse.hpp:
```diff
@@ -0,0 +1,457 @@
+  ProgenitorProfile(const std::string& filename);
+  // interpolate function
+  std::array<double, 6> interpolate(double target_radius,
+                                    double target_cos_theta,  // not needed?
```
I guess decide what to do with the `target_cos_theta` parameter?

pajkosmi replied:
I have decided to keep it, in case we map in multiD progenitors in the future.

pajkosmi replied:
FYI I have addressed the changes, but need to figure out why the CCSN test fails on some clang/gcc builds.  I will squash, rebase and let you know once I track down the crash.

nilsdeppe replied:
Sounds good on both fronts!

nilsdeppe commented on src/PointwiseFunctions/AnalyticData/GrMhd/CcsnCollapse.hpp:
```diff
@@ -0,0 +1,457 @@
+
+  // Begin constants to translate units from cgs to c=G=M_Sun=(k_Boltzmann=1)=1
+  // Divide cm/s by speed_of_light_cgs_ to get into c=G=M_Sun=1 units
+  double speed_of_light_cgs_ = 2.99792458e10;
```
You can mark these `constexpr` (they are compile-time constants)

nilsdeppe commented on src/PointwiseFunctions/AnalyticData/GrMhd/CcsnCollapse.hpp:
```diff
@@ -0,0 +1,457 @@
+  double c2g_dens_ = square(g_times_m_sun_) * newtons_grav_constant_ /
+                     pow<6>(speed_of_light_cgs_);
+
+  // Multiply Kelvin by c2g_temp_ to get into c=G=M_Sun=k_Boltzmann=1 units.
```
Do we actually get to set 4 constants to 1? I thought we only get to set 3 of them to one, but for EOS's you choose M_sun=1.

pajkosmi replied:
We do set Msun to 1, independent of EOS. This helps save us having to redimensionalize a different way for differently formulated EOSs (e.g., tabulated vs hybrid).  Note made in code comments.

nilsdeppe commented on src/PointwiseFunctions/AnalyticData/GrMhd/CcsnCollapse.hpp:
```diff
@@ -0,0 +1,457 @@
+ * readin, setting $c = G = M_\odot = 1$.
+ *
+ * Rotation:
+ * Once the stationary progenitor is constructed, an artificial
```
do you need a blank line here (` *`) to get a line break in the dox?

pajkosmi replied:
Yes, without the extra line it appears as a single paragraph (at least in Google Chrome).

nilsdeppe replied:
ping on the extra line

nilsdeppe commented on src/PointwiseFunctions/AnalyticData/GrMhd/CcsnCollapse.hpp:
```diff
@@ -0,0 +1,457 @@
+ *   * radius, rest mass density, specific internal energy, radial velocity,
+ *     electron fraction, $X$, $\Phi$.
+ *
+ * These columns assume a metric of the form $ds^2 = -e^{2\Phi} dt^2 +
```
I'm not sure if you've looked at the rendered dox, but `\f$` is normally what we need to do to get latex to render, but maybe that's been changed.

pajkosmi replied:
The `$` rendered for me, but I switched to `\f$` for consistency.

nilsdeppe replied:
Huh, neat! Thanks for checking, letting me know, and for making the consistency change :D

nilsdeppe commented on src/PointwiseFunctions/AnalyticData/GrMhd/CcsnCollapse.hpp:
```diff
@@ -0,0 +1,457 @@
+    static type lower_bound() { return 0.; }
+  };
+
+  /// Adiabatic index of the system at readin.
```
add a `///` empty line after this and mark this line with `/// \brief Ad....`

nilsdeppe replied:
ping here

nilsdeppe commented on src/PointwiseFunctions/AnalyticData/GrMhd/CcsnCollapse.hpp:
```diff
@@ -0,0 +1,457 @@
+    static type lower_bound() { return -1; }
+  };
+
+  /// Differential rotation parameter for artificially assigned
```
Is there any guidance you can give either here or maybe in the general documentation on what are interesting values? Or maybe point to some papers that give interesting simulations/values?

pajkosmi replied:
Appropriate references have been added.

nilsdeppe commented on src/PointwiseFunctions/AnalyticData/GrMhd/CcsnCollapse.cpp:
```diff
@@ -0,0 +1,909 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
```
include the `CcsnCollapse.hpp` header first

nilsdeppe commented on src/PointwiseFunctions/AnalyticData/GrMhd/CcsnCollapse.cpp:
```diff
@@ -0,0 +1,909 @@
+namespace detail {
+ProgenitorProfile::ProgenitorProfile(const std::string& filename) {
+  std::ifstream prog_file(filename);
+  std::string header_line;
```
It looks like you forgot to commit your tests

nilsdeppe commented on src/PointwiseFunctions/AnalyticData/GrMhd/CcsnCollapse.cpp:
```diff
@@ -0,0 +1,909 @@
+  using std::sin;
+  // Omega = omega0/(1 + (r/A)^2)
+  // vphi = distance_from_rotation_axis * Omega
+  auto rotational_velocity =
```
`auto` is probably not what you want. With expression templates `auto` will be the expression class of the RHS, not the resulting value. E.g. if you did `auto c = a + b;` then the type would be `blaze::AddExpression<Vector, Vector>` and then every time you use `c` it will do the math, rather than doing the math only once. The flip side is you also want to minimize (or really eliminate) allocations. So what  we often do is something like `get<2>(spatial_velocity) = /*math for rotational velocity*/;`, then use `get<2>(spatial_velocity)` where you need the rotational velocity for the x, y components, and then overwrite `2` at the end. What components/allocations you can reuse depends on the exact calculation, and sometimes then you end up with say `get<2>(spatial_velocity) *= cool_math;`, but this ultimately makes a difference in heavily multithreaded environments (i.e. everywhere nowadays...) since it eliminates system calls and so a serial bottleneck.

nilsdeppe commented on src/PointwiseFunctions/AnalyticData/GrMhd/CcsnCollapse.cpp:
```diff
@@ -0,0 +1,909 @@
+
+  auto lorentz_factor = make_with_value<Scalar<DataType>>(get<0>(x), 0.0);
+  // Compute spatial proper velocity, then Lorentz factor.
+  const auto spatial_metric =
```
just FYI, here there are allocations too, but we have figured out a good way to fix this (yet) with analytic solutions/data, and it's a low priority since it's _only_ in the analytic prescriptions. So leave as-is :)

nilsdeppe reviewed (changes requested):
Overall looks good! I think you forgot to commit the tests.

Just in case you haven't had a chance to look, we have https://spectre-code.org/contributing_to_spectre.html (everything you've done so far is perfect!). The next steps I would suggest are:
1. squash all the current commits into one
2. add new fixup commits addressing commits. Please let the review click `Resolve`. I like to use the :+1: reaction to track which reviewer comments I've addressed.
3. push fixups (and tests), then ping me and I'll look at the fixups and test.

nilsdeppe commented:
My best guess for  the failure is somewhere you do `size_t - number` and that goes below zero and then it wraps and you try to do a gigantic allocation. Not sure why only on Clang.

nilsdeppe commented on src/PointwiseFunctions/AnalyticData/GrMhd/CcsnCollapse.cpp:
```diff
@@ -576,23 +573,34 @@ CcsnCollapse::variables(
       (1.0 + square(vars->radius * inv_diff_rot_parameter_));
 
+  // speed of light check
+  if (UNLIKELY(max(get<2>(spatial_velocity)) > 1.0)) {
```
Should this be an `ASSERT` or an `ERROR`? The reason I ask is `ERROR` is meant to deal with things like input file problems, while `ASSERT` is more meant to deal with code bugs. `ASSERT` checks disappear in release mode and increase performance, while `ERROR`s don't and so add a performance penalty. I _think_ this seems like it should be an `ASSERT` but up to you!

nilsdeppe commented on src/PointwiseFunctions/AnalyticData/GrMhd/CcsnCollapse.hpp:
```diff
@@ -0,0 +1,472 @@
+  };
+
+  /// Central angular velocity artificially assigned at readin.
+  /// Currently limited by the angular velocity of an object rotating in a
```
`///` here too

nilsdeppe reviewed (commented):
LGTM, just a matter of figuring out the test failure

pajkosmi commented:
@nilsdeppe I fleshed out the test coverage and addressed your comments.  The only failing test was `Unit.Evolution.Systems.GeneralizedHarmonic.DuDt` which I didn't modify.

nilsdeppe commented on tests/Unit/PointwiseFunctions/AnalyticData/GrMhd/Test_CcsnCollapse.cpp:
```diff
@@ -0,0 +1,250 @@
+          *deserialized_option_solution);
+
+#ifdef SPECTRE_DEBUG
+  const std::unique_ptr<evolution::initial_data::InitialData>
```
I'm missing something in terms of the code inside this `ifdef`. I don't see a `CHECK` or `CHECK_THROWS_WITH`. Is one of those supposed to be there? Oh! It's used below on 188. Could you move the create, etc. down there so it's clearer how it's used?

nilsdeppe commented on tests/Unit/PointwiseFunctions/AnalyticData/GrMhd/Test_CcsnCollapse.cpp:
```diff
@@ -0,0 +1,250 @@
+      determinant_and_inverse(spatial_metric).first;
+
+  // Metric must always be greater or equal to 1.0 (i.e. flat space)
+  CHECK(abs(min(get(sqrt_det_spatial_metric_analytic)) - 1.0) >= 1.0e-15);
```
These "approximate" things can be check with:
```cpp
CHECK(min(get(sqrt_det_spatial_metric_analytic)) approx(1.0));
```
same with the 2 below. We also have `CHECK_ITERABLE_APPROX` that can be applied over `DataVector`s, etc. You can also use custom `approx` variants that have larger scales (e.g. 1.0e-9 instead of 1.0e-14)

nilsdeppe reviewed (changes requested):
LGTM, couple minor things. Please squash into one commit and have the commit message be something like:
```
Add analytic data for CCSN profile read from disk

# more info here if you have it
```
First line should be <= 60 chars, 2nd line blank, then additional details if they need to be there. We prefer verby commits, e.g. `Add, Remove, Fix`, etc.
